### ISSUE TITLE
Accuracy drops for tensorflow post integer quantization while exporting trained model

### ISSUE THREAD

**Issue by hardikdava** [ultralytics/yolov5#10164]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Export

### Bug

Hello maintainers of yolov5,  I trained yolov5s model according to guidelines and some experiments. I achieved the good results. My end application is to run model on edegtpu which requires quantized int8 format. I noticed that there is a huge accuracy drops after converting tensorflow saved model to int8 tflite model. Here is some insights from validation. 

```
imgsz:320
batch-size:1
iou-threshold:0.6
conf-threshold:0.001

pytorch:
                 Class     Images  Instances          P          R      mAP50   mAP50-95: 100%|██████████| 191
                   all       1915      11640      0.977      0.969      0.992      0.853
                     0       1915      11083      0.989      0.971      0.993      0.876
                     1       1915        557      0.964      0.968      0.991       0.83

Tensorflow saved_model:
                 Class     Images  Instances          P          R      mAP50   mAP50-95: 100%|██████████| 191
                   all       1915      11640      0.976      0.965      0.992      0.851
                     0       1915      11083       0.99      0.967      0.993      0.875
                     1       1915        557      0.963      0.962      0.991      0.827

Tflite-int8:

                 Class     Images  Instances          P          R      mAP50   mAP50-95: 100%|██████████| 191
                   all       1915      11640      0.979      0.962      0.991      0.799
                     0       1915      11083      0.992      0.963      0.993      0.803
                     1       1915        557      0.965      0.961       0.99      0.794

EdgeTPU:
                 Class     Images  Instances          P          R      mAP50   mAP50-95: 100%|██████████| 191
                   all       1915      11640      0.979      0.963      0.991      0.796
                     0       1915      11083      0.992      0.963      0.993      0.801
                     1       1915        557      0.966      0.962      0.989      0.791

```

The accuracy drop was expected but not so much. Is there any possibility/specific reason to reduce accuracy drop?

### Environment

- model version: YoloV5s 
- image size: 320
- batch-size:20(Training)
- iou-threshold:0.6
- conf-threshold:0.001
- epochs: 600
- auto_anchor: yes

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@hardikdava yes that's correct. TFLite FP16 models return same metrics but TF's INT8 quantization strategy results in mAP drops. This is because TFLite models are running actual INT8 inference, whereas other formats that don't see INT8 mAP drops are because the models are merely stored in INT8 but then cast to FP16 or FP32 for inference, i.e. CoreML.

**Comment by hardikdava** [comment]
Hi @glenn-jocher , is it possible to do quantize aware training?

**Comment by glenn-jocher** [comment]
@hardikdava yes that's a good suggestion, that would resolve most of the accuracy drop in TFLite INT8

**Comment by seboz123** [comment]
Hi,
I did notice, that the accuracy drop is mostly due to `tf.lite.Optimize.DEFAULT`
Any idea why that happens?
Trained with Yolov5s, on my Data, I get metrics like this:
best.pt: mAP 0.378
best-no-optimize.tflite: 0.376
best-optimize.tflite: 0.315
That is nearly 20% accuracy loss, I find that quite a lot.
@glenn-jocher Any input why tflite post training quantization leads to such a big accuracy drop?

**Comment by glenn-jocher** [comment]
@seboz123 already explained above

> @hardikdava yes that's correct. TFLite FP16 models return same metrics but TF's INT8 quantization strategy results in mAP drops. This is because TFLite models are running actual INT8 inference, whereas other formats that don't see INT8 mAP drops are because the models are merely stored in INT8 but then cast to FP16 or FP32 for inference, i.e. CoreML.

**Comment by hardikdava** [comment]
Hello @glenn-jocher and @seboz123 After long research, I found that it is all about activation function. `nn.SiLU` is not performing good after quantization specially post training int8 quantization. However `nn.ReLU` is better compare to 'nn.SiLU'. Check [here](https://blog.tensorflow.org/2020/03/higher-accuracy-on-vision-models-with-efficientnet-lite.html). You can find some statatics on custom dataset below.

```
pytorch model:
                 Class     Images  Instances          P          R      mAP50   mAP50-95: 100%|██████████| 1885/1885 [03:27<00:00,  9.10it/s]
                   all       1885      11592       0.96      0.872       0.92      0.634
                     0       1885      11053      0.955      0.973       0.99      0.756
                     1       1885        539      0.965      0.772       0.85      0.513

Tensorflow saved model:
                 Class     Images  Instances          P          R      mAP50   mAP50-95: 100%|██████████| 1885/1885 [02:42<00:00, 11.58it/s]
                   all       1885      11592      0.958      0.868      0.914      0.625
                     0       1885      11053      0.958      0.968       0.99      0.756
                     1       1885        539      0.958      0.767      0.837      0.495

int8-tflite:
                 Class     Images  Instances          P          R      mAP50   mAP50-95: 100%|██████████| 1885/1885 [06:05<00:00,  5.15it/s]
                   all       1885      11592       0.95      0.867      0.911      0.601
                     0       1885      11053      0.951      0.966      0.989      0.713
                     1       1885        539      0.949      0.767      0.832      0.489

edgetpu tflite:
                 Class     Images  Instances          P          R      mAP50   mAP50-95: 100%|██████████| 1885/1885 [03:21<00:00,  9.37it/s]
                   all       1885      11592      0.954      0.865      0.913        0.6
                     0       1885      11053      0.953      0.966      0.989      0.714
                     1       1885        539      0.955      0.764      0.837      0.486
```


One can change activation function directly from model configuration file (models/yolov5*.yaml) : 
Check models/hub/yolov5s-LeakyReLU.yaml
```
activation: nn.ReLU()  # <----- Conv() activation used throughout entire YOLOv5 model
```

Also add activation function to tf export as well which can support `ReLU` for tensorflow model export. Modify `mode/tf.py` activation function as below.

```
def activations(act=nn.SiLU):
    # Returns TF activation from input PyTorch activation
    if isinstance(act, nn.LeakyReLU):
        return lambda x: keras.activations.relu(x, alpha=0.1)
    elif isinstance(act, nn.ReLU):
        return lambda x: keras.activations.relu(x)
    elif isinstance(act, nn.Hardswish):
        return lambda x: x * tf.nn.relu6(x + 3) * 0.166666667
    elif isinstance(act, (nn.SiLU, SiLU)):
        return lambda x: keras.activations.swish(x)
    else:
        raise Exception(f'no matching TensorFlow activation found for PyTorch activation {act}')

```

**Comment by hardikdava** [comment]
@glenn-jocher please checkout my PR #10615 for this issue

**Comment by glenn-jocher** [comment]
@hardikdava thanks for the detailed analysis and for creating the PR #10615. Your contribution is greatly appreciated. I will review it and provide feedback as soon as possible. Thank you for your valuable insight and effort in improving the YOLOv5 model. Keep up the great work!
